pkg/flow/internal/controller/component.go

```diff
@@ -159,19 +160,30 @@ func getManagedOptions(globals ComponentGlobals, cn *ComponentNode) component.Op
 		prefix = "/" + prefix
 	}
 
+	// We need to generate a globally unique component ID to give to the
+	// component and for use with telemetry data. For everything else (HTTP,
+	// data), we can just use the controller-local ID as those values are
+	// guaranteed to be globally unique.
+	globalID := cn.nodeID
+	if globals.ControllerID != "" {
+		globalID = path.Join(globals.ControllerID, cn.nodeID)
+	}
+
 	wrapped := newWrappedRegisterer()
 	cn.register = wrapped
 	return component.Options{
-		ID:            cn.nodeID,
-		Logger:        log.With(globals.Logger, "component", cn.nodeID),
-		DataPath:      filepath.Join(globals.DataPath, cn.nodeID),
-		OnStateChange: cn.setExports,
+		ID:     globalID,
+		Logger: log.With(globals.Logger, "component", globalID),
 		Registerer: prometheus.WrapRegistererWith(prometheus.Labels{
-			"component_id": cn.nodeID,
+			"component_id": globalID,
 		}, wrapped),
-		Tracer:         wrapTracer(globals.TraceProvider, cn.nodeID),
+		Tracer: wrapTracer(globals.TraceProvider, globalID),
+
+		DataPath:       filepath.Join(globals.DataPath, cn.nodeID),
```

Comment:
Should we also use the globalID here so that different modules have a different DataPath?

Reply:
No, see the comments in flow.Options for DataPath: 

```
// Directory where components can write data. Constructed components will be
// given a subdirectory of DataPath using the local ID of the component.
//
// If running multiple Flow controllers, each controller must have a
// different value for DataPath to prevent components from colliding.
```

Reply:
Basically module loaders don't know what the shared DataPath is, because they _already_ get a unique path for their DataPath, and they would propagate that forward to the controllers they instantiate. 

It's a little confusing, but I couldn't find a better way of handling it right now other than documentation.